### ISSUE TITLE
QuickCheck driver always runs the same test

### ISSUE THREAD

**Issue by rudymatela** [shapr/kudzu#8]
I think there is a bug with Kudzu's QuickCheck driver.  Under certain conditions, it seems to only test for a single test case even when configured to run for more than one.

See the following example:

```haskell
import Kudzu

main :: IO ()
main  =  do
  print =<< testUntilSameQC 6 (\x -> x == (0::Int))
```

The property is obviously incorrect and states that every integer is a zero!

When running the above program, this is what I get:

```haskell
+++ OK, passed 1 test.
+++ OK, passed 1 test.
+++ OK, passed 1 test.
+++ OK, passed 1 test.
+++ OK, passed 1 test.
+++ OK, passed 1 test.
+++ OK, passed 1 test.
(6,Just 0)
```

Meaning the tests have passed?

Investigating further with `traceShow`, we can see the reason.  The only test argument across all runs is 0:

```haskell
import Kudzu
import Debug.Trace

prop_zero :: Int -> Bool
prop_zero x | trace ("x = " ++ show x ++ "\n") False  =  undefined
prop_zero x  =  x == 0

main :: IO ()
main  =  do
  print =<< testUntilSameQC 6 prop_zero
```

```haskell
x = 0sts)

+++ OK, passed 1 test.
x = 0sts)

+++ OK, passed 1 test.
x = 0sts)

+++ OK, passed 1 test.
x = 0sts)

+++ OK, passed 1 test.
x = 0sts)

+++ OK, passed 1 test.
x = 0sts)

+++ OK, passed 1 test.
x = 0sts)

+++ OK, passed 1 test.
(6,Just 0)
```

Testing with "pure" QuickCheck hints at the underlying reason:

```haskell
import Test.QuickCheck

main :: IO ()
main  =  do
  quickCheck (withMaxSuccess 1 (\x -> (x::Int) == 0))
  quickCheck (withMaxSuccess 1 (\x -> (x::Int) == 0))
  quickCheck (withMaxSuccess 1 (\x -> (x::Int) == 0))
  quickCheck (withMaxSuccess 1 (\x -> (x::Int) == 0))
  quickCheck (withMaxSuccess 1 (\x -> (x::Int) == 0))
  quickCheck (withMaxSuccess 1 (\x -> (x::Int) == 0))
```

```haskell
+++ OK, passed 1 test.
+++ OK, passed 1 test.
+++ OK, passed 1 test.
+++ OK, passed 1 test.
+++ OK, passed 1 test.
+++ OK, passed 1 test.
```

For a property with an `Int` argument.  When run for just a single test, QuickCheck always tests with the argument "0".  This may be due to QuickCheck trying random small test cases first then progressively bigger cases.  With 1 case, at least for `Int` arguments, there seems to be no randomness.

Maybe Kudzu ought to call QuickCheck in chunks of a dozen or so `maxSuccesses` or use a more low-level internal interface to QuickCheck.  :thinking: :thought_balloon: 

### System information

Here are the versions used in reproducing the bug:

* GHC 9.0.2
* Kudzu's git main head: 0869ea1967062d64ce1a962401aa604e59b917ee
* QuickCheck 2.14.3

**Comment by rudymatela** [comment]
The same behaviour seems to happen for lists, see:

```haskell
import Kudzu
import Debug.Trace

main :: IO ()
main  =  do
  print =<< testUntilSameQC 360 (\xs -> xs == ([] :: [Int]))
```

```haskell
+++ OK, passed 1 test.
+++ OK, passed 1 test.
... (357 lines omitted) ...
+++ OK, passed 1 test.
+++ OK, passed 1 test.
(360,Just 0)
```

**Comment by shapr** [comment]
Oh dear, I'm glad you found this.

**Comment by Rewbert** [comment]
Hi! I am a PhD. student and I work with Koen & John, and have recently done some work together with Nick (QC maintainer) on parallelizing QC, so I am knee-deep in the code right now. I can shed some light on this. @rudymatela you are right in your conclusions of what the root of this issue is.

This is due to how QC generates test data! QC has a size parameter that tries to help you generate small terms first, and then larger ones. Initially, the size is always zero. This size increases as you pass more tests (and after you've discarded a magic number of 10 tests). Eventually it wraps around, etc.

Some generators, e.g the one for lists, starts by generating the length of the list. When the size is zero there can only ever be an empty list generated. Since this size parameter is reset every time you invoke QC, Kudzu will always use the same size (zero). Similarly for `Int`s, the generator for integers will generate numbers within a range, and when the size is zero the range is `(-0, 0)`, so it can only ever generate a zero! That is why your first property holds, because the generated int is always zero in this case. If you had chosen any number `/= 0`, the property would always be falsified instead. **Please note that in these cases it does not matter which random seed you use**!

The thing here is that QC is not really meant to be rerun like this to get fine-grained stats of how it performs. It is meant to run a bunch of tests at once. You could get around this is a couple of ways.
+ If you could run more tests before querying the coverage, the size would vary and eventually falsify your first property. This is by no means a proper solution though, because the next time you invoke QC you will run tests using the same sizes as before. You would never explore the values that can be generated by large sizes this way. Running 10 tests twice (assuming all tests pass) would run tests with sizes `[0..9] ++ [0..9]`, whereas running 20 tests straight away would use sizes `[0..19]`.
+ If QC would let you specify the initial size, you could have rerun it with incrementing sizes and observe the expected behavior. The difficulty with this, currently, is that the seed is not a single counter that is manipulated. It is computed from the number of tests passed so far, and the number of tests that have been discarded since the last non-discarded test. It would take some more engineering work to the code base to actually get this working (meaning, a quick hack is not enough). You can replay a test case (a pair of a seed and a size), so perhaps you can feed in your own seed an an incrementing size? If you can create your own `QCGen` this might actually be a workaround... I don't remember now, but if `Test.QuickCheck.Random` is exposed, you can call `newQCGen` and get a seed that you can feed in through the `stdArgs`, together with a size of your choice.

This size computation strategy is a thorn in my side, and I hope that a master student will want to do a thesis where they evaluate different strategies for this! I will post a call for this at Chalmers where I work.

**Comment by shapr** [comment]
This is fixed by #12
